### DISCUSSION TITLE
Documentation Feedback

### DISCUSSION THREAD

**Discussion by DrRuhe** [winnow-rs/winnow#196] [Ideas]
So, the documentation has left me a bit puzzled. 
Chapter 3 introduces take_while in the context of sequences, without explaining it.
I assumed it meant the given parser would be applied many times.
Therefore I thought `take_while0("abc")` would parse repeating "abc".

Therefore I  believe that the Docs should:
- introduce many0() as the way to repeatedly apply parsers. 
- mention take_while using its argument as a "predicate" that determines if the next character is valid. 
- In the docs for take_while the term "pattern" is used, however, I felt that the terminology pattern implies a string literal to be used as a parser and as a set of possible characters.

Let me know what you think :)

On a side note, the "Ideas" discussions is probably not the best place to put this, but there is no "Feedback" or "General Discussion" section, so here we are.

**Comment by epage** [comment]
Thanks for looking through the tutorial and providing feedback!

The way forward with this feedback is not yet clear.  Chapter 3 didn't explain what the code was doing as it was building off of the example from Chapter 2 which introduced `take_while1`

Also, `many0` and repetition are later covered in Chapter 6.

**Comment by nskobelevs** [comment]
While an old discussion, seems like the most appropriate place to share my own feedback for the current state of the tutorial (as of `0.6.24`).
(I did come across #571 which is more recent but a discussion seems more appropriate for general doc improvements conversation than an issue)

Overall the tutorial is quite good and was great starting point to get a rough idea of winnow but it  wasn't without friction.

## Chapter 7
The main issue for me was chapter 7 on error reporting as it doesn't seem to play well with the previous chapters.

### `parse_digits` changes
While the chapter well demonstrates usage of `Parser::context` the snippets for `parse_digits` feel more like standalone examples than like they're building on the previous chapters.

1. The parser no longer discards the prefix or parses the digits into a `usize`, having a return type of `PResult<(&str, &str)>` when `Hex::from_str` relies on a `PResult<usize>`.
2. The implementation itself is completely different, using `alt` instead of `dispatch!` as was shown in chapter 4.
   - As far as I understand the reason for using `alt` is because in the case of an invalid prefix the 2 chars will still be consumed by `dispatch!(take(2);...` and the error location would be inaccurate if `peek` isn't used. This wasn't entirely obvious though from the start though as this footgun isn't explicitly called out.
   Chapter 2 _does_ have a note of _"[peek](https://docs.rs/winnow/0.6.24/winnow/combinator/fn.peek.html) may be useful when [dispatch](https://docs.rs/winnow/0.6.24/winnow/combinator/macro.dispatch.html)ing from hints from each case’s parser."_ which seems to relate to this but it doesn't fully explain _why_ and to me the wording of that note was hard to properly parse.

### [more minor] `anyhow::Result` comparison with `&str`
```rust
assert_eq!(input.parse::<Hex>().unwrap_err(), error);
```

From Chapter 6 `<Hex as FromStr>::Err` is `anyhow::Error` which can't be compared to a `&str` so presumably `.to_string()` is just missing here.


## General Things
Otherwise I think the tutorial could do with an end goal of "we will build a parser to parse X" which affects the flow here and makes some parts feel disconnected / disjointed. From parsing for non hex digit that is never utilises, to the repetition chapter that feels out of place as it's never referenced to again.

I think it could be beneficial is the tutorial had a goal for a specific parser that would be achieved by the end. If staying on the same theme of parsing number literals even something like parsing a comma separated list of decimal, binary, octal or hexadecimal number, maybe even one that could be nested to have a natural place for the repetition chapter. Ideally something that would give a good introduction to various common parsers but the current state already does that pretty well (although would be nice to include `seq!` somewhere)

I'm also not sure if the imperative style parsers are that useful here. It seems like the declarative approach is most appropriate in the majority of cases so the imperative snippets feels like examples of how not to do things. (I may be partially bias here coming from nom and already being familiar the combinator parsers)

Lastly, most minor, the debugging chapter _could_ do with including a little blurb on interpreting the output of `trace`.


Lemme know what you think!

**Reply by epage** [reply]
imo there is so many points here, it seems worthwhile to make rt a discussion on its own, rather than piggy backing off of someone else's

- We can have a thread per topic, making it easier to follow
- This conversation stays on topic for this specific documentation feedback

**Reply by nskobelevs** [reply]
Makes sense - I've moved the above to #698!
